sphinx/domains/std.py

```diff
@@ -782,11 +782,15 @@ def _resolve_doc_xref(self, env, fromdocname, builder, typ, target, node, contno
         if docname not in env.all_docs:
             return None
         else:
+            title = clean_astext(env.titles[docname])
             if node['refexplicit']:
                 # reference with explicit title
                 caption = node.astext()
+                if '{name}' in caption or '{number}' in caption:
```

Comment:
I think there are three options here:

1. Add substitutions to `:doc:`, as in this patch. This gives cause to wonder if `:ref:` should be enhanced similarly, which suggests that `:numref:` could be deprecated.
2. Introduce a new command that is to `:doc:` as `:numref:` is to `:ref:`. But this requires more logic vs. the first approach for (IMHO) dubious benefit.
3. Teach `:numref:` to also understand document names. But this might be the most difficult.

At any rate, I'm open to suggestions. What the substitution keywords are named is mostly unimportant, except that using the same ones as `:numref:` is less for users to learn/remember. (An alternative I would be okay with, if `:numref:` was also taught to handle it for section references, would be to use `{secnumber}`.)

Reply:
Hmm? Only my concern is about substitution keywords of `:doc:` role.
The target of `:numref:` is a numbered object (ex. figure, table, section and so on). So `{number}` is understandable to me.
On the other hand, the target of `:doc:` is a document, not a section. So I feel `{number}` might not be good for `:doc:` a little.

But I agree using same keywords is easy to learn.

Reply:
Ah, I see. Well... the normal textual resolution of `:doc:` is the document title. And, at least when it appeared in a numbered `toctree` via the `index.rst`, that *does* have a number. It is specifically that number that I want from `{number}`. Thus, FWIW, it makes perfect sense to me, for my use case 🙂.

My original use case — part of it — is that I have a bunch of appendices, e.g. "Appendix I: Foo", "Appendix II: Bar", and I want to write a reference that displays as "Appendix II" rather than "Bar". (The appendix numbers are assigned via a numbered `:toctree:` so I don't need to change them if the order or number of appendices changes, and of course, I want the references to also be automatically in sync.)

(The other part is where I want the *full* page name, e.g. "Appendix I: Foo" rather than "Foo", which I can get with ``:doc:`{number}: {name} <appendix-foo>` ``.)